### ISSUE TITLE
Symbolic locks analysis doesn't handle lock/unlock wrapper functions

### ISSUE THREAD

**Issue by Medowhill** [goblint/analyzer#430]
Hi. Thanks for the great analyzer! I'm working on a research project translating C code with `pthread_mutex` to Rust code with `Mutex`, which couples a mutex and data to protect. For this purpose, I need to find which data are protected by each mutex, and this analyzer seems to fit this goal.

While experimenting with the analyzer, I found a situation in which it loses its precision. Some projects define their own lock/unlock functions calling `pthread_mutex_lock`/`pthread_mutex_unlock`. For example, the following code defines `lock` and `unlock`.

```c
typedef struct { int myint; pthread_mutex_t mymutex; } mystruct;

void lock(mystruct *s) { pthread_mutex_lock(&s->mymutex); }
void unlock(mystruct *s) { pthread_mutex_unlock(&s->mymutex); }

void *foo(void *arg) {
  mystruct *s = (mystruct *) arg;

  lock(s);
  s->myint=s->myint+1;
  unlock(s);

  return NULL;
}

int main() {
  mystruct *s = (mystruct *) malloc(sizeof(*s));
  pthread_mutex_init(&s->mymutex, NULL);

  pthread_t id1, id2;
  pthread_create(&id1, NULL, foo, s);
  pthread_create(&id2, NULL, foo, s);
  pthread_join(id1, NULL);
  pthread_join(id2, NULL);
}
```

When analyzing this code, Goblint reports a data race:

```
[Warning][Race] Memory location (alloc@sid:20).myint@:-1:-1 (race with conf. 110):
  read with {thread:([foo@a.c:23:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  read with {thread:([foo@a.c:24:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  write with {thread:([foo@a.c:23:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  write with {thread:([foo@a.c:24:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
```

However, when `foo` calls `pthread_mutex_lock`/`pthread_mutex_unlock` instead, every access is considered safe as expected.

```c
void *foo(void *arg) {
  mystruct *s = (mystruct *) arg;

  pthread_mutex_lock(&s->mymutex);
  s->myint=s->myint+1;
  pthread_mutex_unlock(&s->mymutex);

  return NULL;
}
```

```
[Success][Race] Memory location (alloc@sid:20).myint@:-1:-1 (safe):
  read with {p-lock:*.mymutex, thread:([foo@a.c:23:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  read with {p-lock:*.mymutex, thread:([foo@a.c:24:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  write with {p-lock:*.mymutex, thread:([foo@a.c:23:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
  write with {p-lock:*.mymutex, thread:([foo@a.c:24:3, main], {})} (conf. 110)  (exp: & s->myint) (a.c:12:3)
```

Is there an option to prevent this imprecision? Currently, I'm using the following configuration:

```json
{
  "ana": {
    "int": {
      "def_exc": true,
      "interval": false,
      "enums": true
    },
    "activated": [
      "expRelation",
      "base",
      "threadid",
      "threadflag",
      "threadreturn",
      "escape",
      "mutex",
      "mallocWrapper",
      "var_eq",
      "symb_locks",
      "region",
      "thread"
    ],
    "context": {
      "widen": true
    }
  },
  "exp": {
    "privatization": "protection"
  }
}
```
I expect I'll ask other questions in the near future. Please let me know if this issue tracker isn't a proper place to ask questions like this.

Thank you!

**Comment by sim642** [comment]
Custom lock/unlock wrapper functions do have some support, e.g. in this test: https://github.com/goblint/analyzer/blob/2bea635e01eab19fe9200d723a547be702fff023/tests/regression/04-mutex/05-lockfuns.c.
Although that's just with a global mutex.

The successful output from your second example shows it's verified it using symbolic locks, which is a somewhat separate mechanism and maybe doesn't support lock/unlock wrappers.

Another related issue could be #186, where the mutex is dynamically allocated like in your examples. Although that might not be the root cause here since without the wrapper functions it does work.

**Comment by sim642** [comment]
> The successful output from your second example shows it's verified it using symbolic locks, which is a somewhat separate mechanism and maybe doesn't support lock/unlock wrappers.

Seems like this is the case indeed. I've suggested a fix in PR #431, which at least seems to fix your example.

**Comment by Medowhill** [comment]
Thanks a lot for the quick response!
